### PR TITLE
Update VecNormalize normalization

### LINKED ISSUE #602: [proposal] Public `VecNormalize._normalize_observation`
I have saved demonstrations from a particular environment, and also a PPO2 policy trained on a normalized version of the same environment.

To properly compare the unnormalized observations in the demonstration against the PPO2 policy I am using `VecNormalize._normalize_observation(obs_demo)`. For ease of use, I'm wondering if the maintainers would be fine with:

1. Rename `_normalize_obs` to `normalize_obs` so that it is a "public" method.
2. Add `training` param to `normalize_obs()` so that we can call the `normalize_obs` without changing normalizing statistics. `VecNormalize.step()` would call `self.normalize_obs(obs, training=self.training)`.

### PR DESCRIPTION
**Pull request by shwang**
## Description

Decouples normalization from mean/stddev tracking functionality.
Exposes reward and observation normalization for use outside of `.step()` and `reset()`.
Add `get_original_rews()` function which returns the most recent unnormalized rewards.
`reset()` now updates statistics for observations in addition to normalizing.

## Motivation and Context

closes #602.
- [x] I have raised an issue to propose this change ([required](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the [changelog](https://github.com/hill-a/stable-baselines/blob/master/docs/misc/changelog.rst) accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have ensured `pytest` and `pytype` both pass.

### CODE REVIEW

**Comment by araffin** [comment]
Thanks for the PR, I'll try to do a review before next week ;)

File: docs/misc/changelog.rst [hunk 1]
```diff
@@ -24,7 +24,14 @@ New Features:
 - Environments are automatically wrapped in a `DummyVecEnv` if needed when passing them to the model constructor
 - Added `stable_baselines.common.make_vec_env` helper to simplify VecEnv creation
 - Added `stable_baselines.common.evaluation.evaluate_policy` helper to simplify model evaluation
-- `VecNormalize` now supports being pickled and unpickled.
+- `VecNormalize` changes:
+
+   - Now supports being pickled and unpickled (@AdamGleave).
+   - New methods `.normalize_observations(obs)` and `normalize_rewards(rews)` apply normalization
+     to arbitrary observation or rewards without updating statistics (@shwang)
+   - `.get_original_rews()` returns the unnormalized rewards from the most recent timestep
```
Comment (AdamGleave) [inline docs/misc/changelog.rst]:
```suggestion
   - `.get_original_reward()` returns the unnormalized rewards from the most recent timestep
```

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 2]
```diff
@@ -88,48 +89,76 @@ def step_wait(self):
...
             if self.training:
-                self.ret_rms.update(self.ret)
-            rews = np.clip(rews / np.sqrt(self.ret_rms.var + self.epsilon), -self.clip_reward, self.clip_reward)
+                self._update_reward(rews)
+            rews = self.normalize_reward(rews)
         self.ret[news] = 0
         return obs, rews, news, infos
 
-    def _normalize_observation(self, obs):
+    def _update_reward(self, rews):
```
Comment (AdamGleave) [inline stable_baselines/common/vec_env/vec_normalize.py]: docstring? (I know it's obvious, but one sentence is still worth it.)

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 3]
```diff
@@ -88,48 +89,76 @@ def step_wait(self):
...
+        self.ret = self.ret * self.gamma + rews
+        self.ret_rms.update(self.ret)
+
+    def normalize_obs(self, obs):
         """
-        :param obs: (numpy tensor)
+        Normalize observations using this VecNormalize's observations statistics.
+        Calling this method does not update statistics.
+
+        :param obs: (numpy float)
```
Comment (AdamGleave) [inline stable_baselines/common/vec_env/vec_normalize.py]: can we make these type annotations? We support this now in SB.

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 4]
```diff
@@ -88,48 +89,76 @@ def step_wait(self):
...
+        obs = np.clip((obs - self.obs_rms.mean) / np.sqrt(self.obs_rms.var + self.epsilon), -self.clip_obs,
+                      self.clip_obs)
+        return obs
+
+    def normalize_reward(self, rews):
+        """
+        Normalize rewards using this VecNormalize's rewards statistics.
+        Calling this method does not update statistics.
+
+        :param rews: (numpy float)
```
Comment (AdamGleave) [inline stable_baselines/common/vec_env/vec_normalize.py]: same -- type annotations

File: tests/test_vec_normalize.py [hunk 5]
```diff
@@ -76,6 +76,42 @@ def test_vec_env(tmpdir):
     check_vec_norm_equal(norm_venv, deserialized)
 
 
+def _make_stepped_cartpole():
```
Comment (AdamGleave) [inline tests/test_vec_normalize.py]: Why do we need this? To ensure the normalization statistics are "warm-started"? Good to add a docstring explaining this.

File: tests/test_vec_normalize.py [hunk 6]
```diff
@@ -76,6 +76,42 @@ def test_vec_env(tmpdir):
...
+    return venv
+
+
+def test_get_original():
+    venv = _make_stepped_cartpole()
+    acts = [venv.action_space.sample()]
+    obs, rews, _, _ = venv.step(acts)
+    orig_obs = venv.get_original_obs()
+    orig_rews = venv.get_original_reward()
+    assert not np.array_equal(orig_obs, obs)
```
Comment (AdamGleave) [inline tests/test_vec_normalize.py]: Can we add some more checks here? Ideas:
  - shape and type of `orig_obs` and `obs` should match
  - does `venv.normalize_obs(orig_obs)` match `obs` and ditto for reward?

Comment (shwang) [reply tests/test_vec_normalize.py]: Added these checks and also loop the normalization checks over 3 steps of CartPole now.

File: stable_baselines/common/vec_env/vec_normalize.py (hunk 3 above)
Comment (shwang) [reply stable_baselines/common/vec_env/vec_normalize.py]: done, I can also annotate the rest of the file if that's helpful

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 7]
```diff
@@ -88,48 +89,67 @@ def step_wait(self):
...
-                self.ret_rms.update(self.ret)
-            rews = np.clip(rews / np.sqrt(self.ret_rms.var + self.epsilon), -self.clip_reward, self.clip_reward)
+                self._update_reward(rews)
+            rews = self.normalize_reward(rews)
         self.ret[news] = 0
         return obs, rews, news, infos
 
-    def _normalize_observation(self, obs):
+    def _update_reward(self, rews: np.ndarray) -> None:
+        """Update reward normalizing statistics."""
```
Comment (AdamGleave) [inline stable_baselines/common/vec_env/vec_normalize.py]:
```suggestion
        """Update reward normalization statistics."""
```

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 8]
```diff
@@ -88,48 +89,67 @@ def step_wait(self):
...
         """
-        if self.norm_obs:
-            if self.training:
-                self.obs_rms.update(obs)
-            obs = np.clip((obs - self.obs_rms.mean) / np.sqrt(self.obs_rms.var + self.epsilon), -self.clip_obs,
-                          self.clip_obs)
-            return obs
-        else:
-            return obs
+        obs = np.clip((obs - self.obs_rms.mean) / np.sqrt(self.obs_rms.var + self.epsilon), -self.clip_obs,
```
Comment (araffin) [inline stable_baselines/common/vec_env/vec_normalize.py]: @AdamGleave Should we add a check like:
```python
if self.norm_obs:
     ...
return obs
```
?

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 9]
```diff
@@ -88,48 +89,67 @@ def step_wait(self):
...
     def reset(self):
         """
         Reset all environments
         """
         obs = self.venv.reset()
-        if len(np.array(obs).shape) == 1:  # for when num_cpu is 1
-            self.old_obs = [obs]
-        else:
-            self.old_obs = obs
+        self.old_obs = obs
```
Comment (araffin) [inline stable_baselines/common/vec_env/vec_normalize.py]: @AdamGleave Shouldn't it be `obs.copy` to avoid modifying `self.old_obs` because here both variables point to the same object?

File: tests/test_vec_normalize.py [hunk 10]
```diff
@@ -76,6 +76,53 @@ def test_vec_env(tmpdir):
...
+    for _ in range(100):
+        acts = [venv.action_space.sample()]
+        venv.step(acts)
+    return venv
+
+
+def test_get_original():
+    venv = _make_warmstart_cartpole()
+    for _ in range(3):
+        acts = [venv.action_space.sample()]
```
Comment (araffin) [inline tests/test_vec_normalize.py]: I know the old code uses `acts` and `rews` but please use `actions` and  `rewards`,
it is not that much longer but it is clearer

File: stable_baselines/common/vec_env/vec_normalize.py (hunk 8 above)
Comment (AdamGleave) [reply stable_baselines/common/vec_env/vec_normalize.py]: In favour of this (and similarly in `normalize_reward`). In that case I think we should also update callees to no longer check `self.norm_obs`.

File: stable_baselines/common/vec_env/vec_normalize.py (hunk 9 above)
Comment (AdamGleave) [reply stable_baselines/common/vec_env/vec_normalize.py]: Copying here would be defensive but I don't think is necessary. `self.old_obs` should never get modified?

Comment (araffin) [reply stable_baselines/common/vec_env/vec_normalize.py]: true, if was more afraid of `obs` getting modified outside the wrapper, but I don't remember in which case I had a problem (I think it was with a replay buffer)

File: stable_baselines/common/vec_env/vec_normalize.py (hunk 8 above)
Comment (shwang) [reply stable_baselines/common/vec_env/vec_normalize.py]: done

Comment (shwang) [reply stable_baselines/common/vec_env/vec_normalize.py]: done

**Comment by shwang** [comment]
Thanks for the reviews!

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 11]
```diff
@@ -88,48 +89,69 @@ def step_wait(self):
...
-        self.ret = self.ret * self.gamma + rews
         self.old_obs = obs
-        obs = self._normalize_observation(obs)
-        if self.norm_reward:
-            if self.training:
-                self.ret_rms.update(self.ret)
-            rews = np.clip(rews / np.sqrt(self.ret_rms.var + self.epsilon), -self.clip_reward, self.clip_reward)
+        self.old_rews = rews
+
+        if self.training:
```
Comment (AdamGleave) [inline stable_baselines/common/vec_env/vec_normalize.py]: @araffin do you think we should update observation statistics even if `self.norm_obs` is false? It might have a slight resource overhead but otherwise can't see a disadvantage.

File: stable_baselines/common/vec_env/vec_normalize.py [hunk 12]
```diff
@@ -88,48 +89,69 @@ def step_wait(self):
...
+                           -self.clip_reward, self.clip_reward)
+        return reward
 
-        :return: (numpy float)
+    def get_original_obs(self) -> np.ndarray:
+        """
+        Returns an unnormalized version of the observations from the most recent
+        step or reset.
         """
         return self.old_obs
```
Comment (AdamGleave) [inline stable_baselines/common/vec_env/vec_normalize.py]: @araffin do you think we should do `self.old_obs.copy()` here defensively (related to your previous comment)? If so same thing in `get_original_reward`.

**Comment by araffin** [comment]
i dont have a strong opinions for your two comments

File: stable_baselines/common/vec_env/vec_normalize.py (hunk 12 above)
Comment (AdamGleave) [reply stable_baselines/common/vec_env/vec_normalize.py]: I've opted to do a defensive copy here since this seems like it could lead to hard to track bugs.

File: stable_baselines/common/vec_env/vec_normalize.py (hunk 11 above)
Comment (AdamGleave) [reply stable_baselines/common/vec_env/vec_normalize.py]: Decided against this. Don't have strong feelings either way but the resource overhead seems minimal and I could imagine use cases where this is actually helpful (e.g. running two policies, one of which wants normalized and other not, and still want to update observation statistics from both).
